vm/vm.go

```diff
@@ -50,10 +38,41 @@ func NewVM(debug bool) *VM {
 	return vm
 }
 
+func (vm *VM) RunSafe(program *Program, env interface{}) (out interface{}, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			h := file.Error{
+				Location: program.Locations[vm.pp],
+				Message:  fmt.Sprintf("%v", r),
+			}
+			err = fmt.Errorf("%v", h.Format(program.Source))
+		}
+	}()
+
+	out = vm.Run(program, env)
+	return
+}
+
+func (vm *VM) Reset() {
+	vm.stack = vm.stack[0:0] // keep the existing memory/capacity
+	vm.ip = 0
+	vm.pp = 0
+	if vm.scopes != nil {
+		vm.scopes = vm.scopes[0:0] // keep the existing memory/capacity
+	}
+}
+
 func (vm *VM) Run(program *Program, env interface{}) interface{} {
 	vm.bytecode = program.Bytecode
 	vm.constants = program.Constants
 
+	// Avoid having to dynamically allocate the calling arguments frame in the
+	// typical case of calls, if larger number of arguments then will alloc.
+	// This is possible because the vm never calls back into itself / nests calls
+	// (expr code cannot call into a function defined in expr code), so the
+	// callstack is only ever a single frame.
+	var fixedFrame20 [20]reflect.Value
```

Comment:
How does it affect expressions without calls at all? We gonna allocate frames for every VM run. Even there is no calls. Mabe this part should be separated in another PR? And added compile time detection if it's needed to allocate fixedFrame20. Also is there any performance boost from it?